src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -68,187 +66,194 @@ ShardManager::Result< ShardInfo > HSHomeObject::_create_shard(pg_id_t pg_owner,
         }
         repl_dev = static_cast< HS_PG* >(iter->second.get())->repl_dev_;
     }
+
     if (!repl_dev) {
         LOGW("failed to get repl dev instance for pg [{}]", pg_owner);
         return folly::makeUnexpected(ShardError::PG_NOT_READY);
     }
 
     auto new_shard_id = generate_new_shard_id(pg_owner);
     auto create_time = get_current_timestamp();
-    auto shard_info =
-        ShardInfo(new_shard_id, pg_owner, ShardInfo::State::OPEN, create_time, create_time, size_bytes, size_bytes, 0);
-    std::string const create_shard_message = serialize_shard(Shard(shard_info));
-
-    // prepare msg header;
-    const uint32_t needed_size = sizeof(ReplicationMessageHeader) + create_shard_message.size();
-    auto req = repl_result_ctx< ShardManager::Result< ShardInfo > >::make(needed_size);
-
-    uint8_t* raw_ptr = req->hdr_buf_.bytes;
-    ReplicationMessageHeader* header = new (raw_ptr) ReplicationMessageHeader();
-    header->message_type = ReplicationMessageType::SHARD_MESSAGE;
-    header->payload_size = create_shard_message.size();
-    header->payload_crc =
-        crc32_ieee(init_crc32, r_cast< const uint8_t* >(create_shard_message.c_str()), create_shard_message.size());
-    header->header_crc = header->calculate_crc();
-    raw_ptr += sizeof(ReplicationMessageHeader);
-    std::memcpy(raw_ptr, create_shard_message.c_str(), create_shard_message.size());
-
+    std::string const create_shard_message = serialize_shard_info(
+        ShardInfo(new_shard_id, pg_owner, ShardInfo::State::OPEN, create_time, create_time, size_bytes, size_bytes, 0));
+    const auto msg_size = sisl::round_up(create_shard_message.size(), repl_dev->get_blk_size());
+    auto req = repl_result_ctx< ShardManager::Result< ShardInfo > >::make(msg_size, 512 /*alignment*/);
+    auto buf_ptr = req->hdr_buf_.bytes;
+    std::memset(buf_ptr, 0, msg_size);
+    std::memcpy(buf_ptr, create_shard_message.c_str(), create_shard_message.size());
+    // preapre msg header;
+    req->header_.msg_type = ReplicationMessageType::CREATE_SHARD_MSG;
+    req->header_.pg_id = pg_owner;
+    req->header_.shard_id = new_shard_id;
+    req->header_.payload_size = msg_size;
+    req->header_.payload_crc = crc32_ieee(init_crc32, buf_ptr, msg_size);
+    req->header_.seal();
+    sisl::blob header;
+    header.bytes = r_cast< uint8_t* >(&req->header_);
+    header.size = sizeof(req->header_);
+    sisl::sg_list value;
+    value.size = msg_size;
+    value.iovs.push_back(iovec(buf_ptr, msg_size));
     // replicate this create shard message to PG members;
-    repl_dev->async_alloc_write(req->hdr_buf_, sisl::blob{}, sisl::sg_list{}, req);
+    repl_dev->async_alloc_write(header, sisl::blob{}, value, req);
     auto info = req->result().get();
```

Comment:
Its blocking. As brian suggested, we need to change to AsyncResult.

Reply:
same as above.